### DISCUSSION TITLE
Provide a sub-command that searches for `.pyz` files

### DISCUSSION THREAD

**Discussion by brettcannon** [brettcannon/python-launcher#221] [Ideas]
https://docs.python.org/3/library/zipapp.html provides a way to create a zip file that Python can directly execute. Unfortunately you have to point Python at the resulting `.pyz` file to execute it. Providing a generic command that can search `PATH` for such `.pyz` files and use them appropriately may be useful.

Idea sparked by https://discuss.python.org/t/pip-plans-to-introduce-an-alternative-zipapp-deployment-method/17431/25 .

This might not be too critical, though, with an appropriate shebang line on the `.pyz` file.

**Comment by brettcannon** [comment]
This would also make for an interesting alternative to [pipx](https://pypi.org/project/pipx/), as projects could distribute (or be made into) a `.pyz` file instead of needing pipx to create a virtual environment (which always runs the risk of becoming invalid).

**Comment by brettcannon** [comment]
An alternative name is `py z` since `run` is extremely generic; `py z pip`. Maybe too cute, though?

**Reply by merwok** [reply]
Definitely too cryptic!

**Reply by brettcannon** [reply]
But what does `py run` do? That's not descriptive either in the grand scheme of Python "running" something.

**Reply by merwok** [reply]
I don’t expect anything to be descriptive out of context, only useful mnemonic.
Run meaning find and execute fits with other language / environment managers.

**Reply by maxwell-k** [reply]
I like it, and the `z` makes me think zipapp immediately, `run` doesn't

**Reply by merwok** [reply]
`py app pip` then!

**Comment by brettcannon** [comment]
Another option is an entirely separate `pyz` binary, but based on how folks have historically asked me to incorporate more functionality into the Launcher rather than less, I suspect this won't be people's preferred option.

**Comment by jaraco** [comment]
Although I commend the brainstorming here, maybe there's a better experience to be had.

Windows handles this situation by having .pyz in the PATHEXT and having .pyz associated with the executable `py`... and then they don't even have to type `py myexe.pyz` but can simply type `myexe`. Perhaps there's a way to bring this simple user experience to Unix as well, either by shell extensions that implement the same behavior or by providing a wrapper for the .pyz file that gives it direct executability on Python (maybe the right shebang header can already do this?).

I'd really like for there to be an easy way for a developer to provide a _command_ to a user on any platform, without the need for a user to explicitly invoke a launcher or interpreter.

**Reply by brettcannon** [reply]
I don't know if such a solution exists unfortunately within Unix shells. I also wouldn't want to own trying to make this work in every shell.

**Comment by brettcannon** [comment]
PEP 723 makes a `py run` more generic and which can easily implicitly support Python zip files.
